### ISSUE TITLE
How to make Fisher Matrix for other models?

### ISSUE THREAD

**Issue by GaengSu** [talreiss/PANDA#3]
Hello :) I'm so insterested in your work from korea. 

In your repo, Fisher matrix you share is limited to resnet only, and I want to try PANDA(PANDA-EWC) with efficientnet and other latest models.
Your paper does not specify how to create a pth file you shared.
Can you tell me how to make or keywords?

Thanks😊

**Comment by NivC** [comment]
Hi, Thanks for your interest! :-)

Attaching below the method we use to calculate the fisher information matrix. I believe you should be able to apply a similar method with efficientnet.

Best,
Niv.

--------------------------------------

```
def estimate_fisher(self, data_loader, sample_size, batch_size=32):
        # Init
        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
        self.old_model = deepcopy(self).to(device)
        self.old_model.eval()
        ResNet.freeze_model(self.old_model)
        fisher = {}
        ce = torch.nn.CrossEntropyLoss()
        for n, p in self.named_parameters():
            fisher[n] = 0 * p.data

        # ----- Freezing layers -------
        self.freeze_parameters()

        # ----- Compute --------
        self.eval()
        for i, data in enumerate(data_loader, 0):
            self.zero_grad()
            if i >= sample_size // batch_size:
                break
            x, y = data
            x = x.to(device)
            y = y.to(device)
            outputs, _ = self(x)
            loss = ce(outputs, y)
            loss.backward()
            for n,p in self.named_parameters():
                if p.grad is not None:
                    fisher[n] += batch_size * p.grad.data.pow(2)
        for n, _ in self.named_parameters():
            fisher[n] /= sample_size
            fisher[n] = Variable(fisher[n], requires_grad=False)
        self.fisher = fisher
        return fisher
```

**Comment by GaengSu** [comment]
@NivC
Ohhhhh thanks for your reply!! 😊😊😊
I have an additional question. 
The EWC technique of Continental Learning was used to prevent Collapse phenomenon, which is not intuitively understood. 
Can you explain it easily?

thank you!

**Comment by NivC** [comment]
You are welcome.

We claim that the collapse phenomenon in our case is a special case of catastrophic forgetting. Therefore, we use EWC regularization to keep the network weights within a reasonable distance from their pretrained value. You can read more about that in the original EWC paper.

Niv.

**Comment by GaengSu** [comment]
> You are welcome.
> 
> We claim that the collapse phenomenon in our case is a special case of catastrophic forgetting. Therefore, we use EWC regularization to keep the network weights within a reasonable distance from their pretrained value. You can read more about that in the original EWC paper.
> 
> Niv.

Oh thanks 😊😊😊

thank you for your reply

**Comment by GaengSu** [comment]
> You are welcome.
> 
> We claim that the collapse phenomenon in our case is a special case of catastrophic forgetting. Therefore, we use EWC regularization to keep the network weights within a reasonable distance from their pretrained value. You can read more about that in the original EWC paper.
> 
> Niv.
@NivC 
ohhh...! can i ask you last question? The performance has improved significantly with EWC, why?

thanks!

**Comment by NivC** [comment]
The improvement when training with the compactness loss is described in much detail in the paper. EWC mitigates the collapse, also discussed there.

Niv.

**Comment by GaengSu** [comment]
> The improvement when training with the compactness loss is described in much detail in the paper. EWC mitigates the collapse, also discussed there.
> 
> Niv.

thanks!

**Comment by GaengSu** [comment]
> The improvement when training with the compactness loss is described in much detail in the paper. EWC mitigates the collapse, also discussed there.
> 
> Niv.

thanks!
